Conflict between several @types package

<!-- BUGS: Please use this template. -->
<!-- QUESTIONS: This is not a general support forum! Ask Qs at http://stackoverflow.com/questions/tagged/typescript -->
<!-- SUGGESTIONS: See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md -->

**TypeScript Version:**  2.0.0

**Details**
I started a new project to try TS 2.0 and the `npm install @types/` way of getting typings and ran into a conflict issue.
```
$ npm install @types/react
$ npm install @types/react-router
```
React-router definition depends on `@types/react`.

**Expected behavior:**
Everything working fine

**Actual behavior:** 
```
ERROR in /home/vincent/Code/mobx-test/node_modules/@types/react-router/node_modules/@types/react/index.d.ts
(7,21): error TS2300: Duplicate identifier 'React'.

ERROR in /home/vincent/Code/mobx-test/node_modules/@types/react/index.d.ts
(7,21): error TS2300: Duplicate identifier 'React'.
```

Here is a related problem and description:
File main.ts:
```
/// <reference path="jquery-1.10.d.ts"/>

import $ from "jquery"
import {myFunction} from "./my-module"

$('#id').click(function () {});
```

File my-module.ts:
```
/// <reference path="jquery-2.0.d.ts"/>

import $ from "jquery"

export function myFunction() {}
```

Files jquery-1.10.d.ts and jquery-2.0.d.ts are files with the same content, just for testing, but it can be slightly different for different version of the jQuery, and theirs content is latest version of the jquery.d.ts from the DefinitelyTyped (https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/master/jquery/jquery.d.ts):
```
 md5sum jquery-1.10.d.ts jquery-2.0.d.ts
7ef0067d2faf006343aa95157f492867  jquery-1.10.d.ts
7ef0067d2faf006343aa95157f492867  jquery-2.0.d.ts
```

Now try to compile the main.ts with:
```
tsc --moduleResolution node -t ES5 main.ts
```

Expected result: **no errors**

Actual result:
```
jquery-2.0.d.ts(615,5): error TS2374: Duplicate string index signature.
jquery-2.0.d.ts(2844,5): error TS2374: Duplicate string index signature.
jquery-2.0.d.ts(2845,5): error TS2375: Duplicate number index signature.
jquery-2.0.d.ts(3218,5): error TS2300: Duplicate identifier 'export='.
my-module.ts(3,8): error TS1192: Module '"jquery"' has no default export.
jquery-1.10.d.ts(3218,5): error TS2300: Duplicate identifier 'export='.
main.ts(3,8): error TS1192: Module '"jquery"' has no default export.
```
I.e. you see batch of errors. It seems like tsc does not allow use different version of the .d.ts files with the same declarations for the included module/file. However I think it should not produce errors because the included module can use different version of the same library as shown above with file names - jquery-1.10.d.ts and jquery-2.0.d.ts.

tsc version:
```
tsc --version
Version 2.1.0-dev.20160805
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -18763,7 +18763,13 @@ namespace ts {
                     (augmentations || (augmentations = [])).push(file.moduleAugmentations);
                 }
                 if (file.symbol && file.symbol.globalExports) {
-                    mergeSymbolTable(globals, file.symbol.globalExports);
+                    // Merge in UMD exports with first-in-wins semantics (see #9771)
+                    const source = file.symbol.globalExports;
+                    for (const id in source) {
+                        if (!(id in globals)) {
+                            globals[id] = source[id];
+                        }
+                    }
                 }
             });
 

--- a/tests/baselines/reference/umdGlobalConflict.js
+++ b/tests/baselines/reference/umdGlobalConflict.js
@@ -1,0 +1,23 @@
+//// [tests/cases/compiler/umdGlobalConflict.ts] ////
+
+//// [index.d.ts]
+export as namespace Alpha;
+export var x: string;
+
+//// [index.d.ts]
+export as namespace Alpha;
+export var y: number;
+
+//// [consumer.ts]
+import * as v1 from './v1';
+import * as v2 from './v2';
+
+//// [global.ts]
+// Should be OK, first in wins
+const p: string = Alpha.x;
+
+//// [consumer.js]
+"use strict";
+//// [global.js]
+// Should be OK, first in wins
+var p = Alpha.x;

--- a/tests/baselines/reference/umdGlobalConflict.symbols
+++ b/tests/baselines/reference/umdGlobalConflict.symbols
@@ -1,0 +1,29 @@
+=== tests/cases/compiler/v1/index.d.ts ===
+export as namespace Alpha;
+>Alpha : Symbol(Alpha, Decl(index.d.ts, 0, 0))
+
+export var x: string;
+>x : Symbol(x, Decl(index.d.ts, 1, 10))
+
+=== tests/cases/compiler/v2/index.d.ts ===
+export as namespace Alpha;
+>Alpha : Symbol(Alpha, Decl(index.d.ts, 0, 0))
+
+export var y: number;
+>y : Symbol(y, Decl(index.d.ts, 1, 10))
+
+=== tests/cases/compiler/consumer.ts ===
+import * as v1 from './v1';
+>v1 : Symbol(v1, Decl(consumer.ts, 0, 6))
+
+import * as v2 from './v2';
+>v2 : Symbol(v2, Decl(consumer.ts, 1, 6))
+
+=== tests/cases/compiler/global.ts ===
+// Should be OK, first in wins
+const p: string = Alpha.x;
+>p : Symbol(p, Decl(global.ts, 1, 5))
+>Alpha.x : Symbol(Alpha.x, Decl(index.d.ts, 1, 10))
+>Alpha : Symbol(Alpha, Decl(index.d.ts, 0, 0))
+>x : Symbol(Alpha.x, Decl(index.d.ts, 1, 10))
+

--- a/tests/baselines/reference/umdGlobalConflict.types
+++ b/tests/baselines/reference/umdGlobalConflict.types
@@ -1,0 +1,29 @@
+=== tests/cases/compiler/v1/index.d.ts ===
+export as namespace Alpha;
+>Alpha : typeof Alpha
+
+export var x: string;
+>x : string
+
+=== tests/cases/compiler/v2/index.d.ts ===
+export as namespace Alpha;
+>Alpha : typeof 
+
+export var y: number;
+>y : number
+
+=== tests/cases/compiler/consumer.ts ===
+import * as v1 from './v1';
+>v1 : typeof v1
+
+import * as v2 from './v2';
+>v2 : typeof v2
+
+=== tests/cases/compiler/global.ts ===
+// Should be OK, first in wins
+const p: string = Alpha.x;
+>p : string
+>Alpha.x : string
+>Alpha : typeof Alpha
+>x : string
+

--- a/tests/cases/compiler/umdGlobalConflict.ts
+++ b/tests/cases/compiler/umdGlobalConflict.ts
@@ -1,0 +1,15 @@
+//@filename: v1/index.d.ts
+export as namespace Alpha;
+export var x: string;
+
+//@filename: v2/index.d.ts
+export as namespace Alpha;
+export var y: number;
+
+//@filename: consumer.ts
+import * as v1 from './v1';
+import * as v2 from './v2';
+
+//@filename: global.ts
+// Should be OK, first in wins
+const p: string = Alpha.x;

